platform/pulp/plugin/changeset/model.py

```diff
@@ -0,0 +1,162 @@
+    contained within the local (pulp) repository.
+
+    Attributes:
+        model (pulp.plugin.Content): A real content model object.
+        settled (bool): Indicates that all artifacts have been downloaded
+            and the content is ready to be imported.
+
+        artifacts (set): The set of related `RemoteArtifact`.
+    """
+
+    def __init__(self, model):
+        """
+        Args:
+            model (pulp.plugin.Content): A real content model object.
+
+        """
+        super(RemoteContent, self).__init__(model)
+        self.artifacts = set()
+        self.settled = False
+
+    @property
+    def key(self):
+        """
+        The natural key as a dictionary.
+
+        Returns:
+            dict: The natural key.
+        """
+        return {f.name: getattr(self.model, f.name) for f in self.model.natural_key_fields}
+
+    def update(self, model):
+        """
+        Update this `model` stored with the specified model that has been
+        fetched from the database. The artifacts are matched by `relative_path`
+        and their model object is replaced by the fetched model.
+
+        Args:
+            model (pulp.plugin.Content): A fetched content model object.
+        """
+        self.model = model
+        known = {a.model.relative_path: a for a in self.artifacts}
+        self.artifacts.clear()
+        for artifact in model.artifacts.all():
+            try:
+                found = known[artifact.relative_path]
+                found.model = artifact
+                self.artifacts.add(found)
+            except KeyError:
+                log.error(_('Artifact not matched.'))
+
+    def settle(self):
+        """
+        Settle matters that are prerequisite to associating the content
+        with a repository.  This mainly pertains to the artifacts being downloaded;
+        and associated artifact models saved; the content model saved.
+
+        Notes:
+            Called whenever an artifact has been downloaded.
+        """
+        for artifact in self.artifacts:
+            if not artifact.model.downloaded:
+                return
+        self.save()
+        self.settled = True
+
+    def save(self):
+        """
+        Save the content and related artifacts in a single DB transaction.
+        Due to race conditions, the content may already exist raising an IntegrityError.
+        When this happens, the model is fetched and replaced.
+        """
+        is_duplicate = False
+        with transaction.atomic():
+            try:
+                super(RemoteContent, self).save()
+            except IntegrityError:
+                is_duplicate = True
+            else:
+                for artifact in self.artifacts:
+                    artifact.model.content = self.model
+                    artifact.save()
+        if is_duplicate:
+            model = type(self.model)
+            content = model.objects.get(**self.key)
+            self.update(content)
```

Comment:
Yes, the query needed to be outside of the transaction or an exception was raised by the django ORM.